tidy3d/plugins/microwave/custom_path_integrals.py

```diff
@@ -0,0 +1,163 @@
+        if v_field_name not in em_field.field_components:
+            raise DataError(f"'field_name' '{v_field_name}' not found.")
+
+        # Select fields lying on the plane
+        plane_indexer = {dim3: self.position}
+        field1 = em_field.field_components[h_field_name].sel(plane_indexer, method="nearest")
+        field2 = em_field.field_components[v_field_name].sel(plane_indexer, method="nearest")
+
+        # Dim 's' represents the parameterization of the line
+        # 't' is likely used for time
+        x_path = xr.DataArray(self.vertices[:, 0], dims="s")
+        y_path = xr.DataArray(self.vertices[:, 1], dims="s")
+        field1_interp = field1.interp(x=x_path, y=y_path, method="linear")
+        field2_interp = field2.interp(x=x_path, y=y_path, method="linear")
+
+        # Determine the differential length elements along the path
+        dl_x = self._compute_dl_component(x_path, self.is_closed_contour)
+        dl_y = self._compute_dl_component(y_path, self.is_closed_contour)
+        dl_x = xr.DataArray(dl_x, dims="s")
+        dl_y = xr.DataArray(dl_y, dims="s")
+
+        # Compute the dot product between differential length element and vector field
+        integrand = field1_interp * dl_x + field2_interp * dl_y
+        # Integrate along the path
+        result = integrand.integrate(coord="s")
+        result = result.reset_coords(drop=True)
+
+        if isinstance(em_field, FieldData):
+            return FreqDataArray(data=result.data, coords=result.coords)
+        elif isinstance(em_field, FieldTimeData):
+            return TimeDataArray(data=result.data, coords=result.coords)
+        else:
+            assert isinstance(em_field, ModeSolverData)
+            return FreqModeDataArray(data=result.data, coords=result.coords)
+
+    @staticmethod
+    def _compute_dl_component(coord_array: xr.DataArray, closed_contour=False) -> np.array:
+        """Computes the differential length element along the integration path."""
+        dl = np.gradient(coord_array)
+        if closed_contour:
+            end_points = (coord_array[1] - coord_array[-2]).values / 2
+            dl[0] = end_points
+            dl[-1] = end_points
+        return dl
+
+    @cached_property
+    def is_closed_contour(self) -> bool:
+        return (self.vertices[0, :] == self.vertices[-1, :]).all()
+
+    @cached_property
+    def main_axis(self) -> Axis:
+        """Axis for performing integration."""
+        return self.axis
+
+    @pd.validator("vertices", always=True)
+    def _correct_shape(cls, val):
+        """Makes sure vertices size is correct."""
+        # overall shape of vertices
+        if val.shape[1] != 2:
+            raise SetupError(
+                "'CustomPathIntegral2D.vertices' must be a 2 dimensional array shaped (N, 2). "
+                f"Given array with shape of '{val.shape}'."
+            )
+        return val
+
+
+class CustomVoltageIntegral2D(CustomPathIntegral2D):
```

Comment:
I think we went with the module approach (vs the class method approach) to these sorts of things? I'm not opposed to the class approach, but what's the reasoning? do we plan to add fields to the CustomPathIntegral classes down the line?

Reply:
The reasoning here is that although these classes are basically just renaming the base class they do have data associated with them, so overall I think they make sense as a class. Originally I did have some additional members but I realized that they were more confusing than helpful. (e.g. I had a sign parameter, but users can just define their path in the opposite direction if they wanted).

Reply:
Also to mirror how the AxisAligned versions of these integrals are used.